src/Microsoft.AspNetCore.Routing/CompositeEndpointDataSource.cs

```diff
@@ -106,5 +111,52 @@ private void CreateChangeToken()
             _cts = new CancellationTokenSource();
             _consumerChangeToken = new CancellationChangeToken(_cts.Token);
         }
+
+        private string DebuggerDisplayString => ToDebuggerDisplayString();
+
+        private string ToDebuggerDisplayString()
+        {
+            // Try using private variable '_endpoints' to avoid initialization
+            if (_endpoints == null)
+            {
+                return "No endpoints available yet to display.";
+            }
+
+            var sb = new StringBuilder();
+            foreach (var endpoint in _endpoints)
+            {
+                if (endpoint is MatcherEndpoint matcherEndpoint)
+                {
+                    var template = matcherEndpoint.RoutePattern.RawText;
+                    template = string.IsNullOrEmpty(template) ? "\"\"" : template;
+                    sb.Append(template);
+                    var requiredValues = matcherEndpoint.RequiredValues.Select(kvp => $"{kvp.Key} = \"{kvp.Value ?? "null"}\"");
+                    sb.Append(", Required Values: new { ");
+                    sb.Append(string.Join(", ", requiredValues));
+                    sb.Append(" }");
+                    sb.Append(", Order:");
+                    sb.Append(matcherEndpoint.Order);
+
+                    var endpointConstraintMetadata = matcherEndpoint.Metadata.GetOrderedMetadata<IEndpointConstraintMetadata>();
+                    foreach (var constraint in endpointConstraintMetadata)
+                    {
+                        if (constraint is HttpMethodEndpointConstraint httpMethodEndpointConstraint)
```

Comment:
What about adding internal virtual DebuggerToString to the constraint and then doing a foreach of the constraints. By default it would be type name + order. Override in HttpMethodEndpointConstraint to print methods.

Reply:
I suppose you thought the base was an abstract type?, its actually an interface (IEndpointConstraintMetadata)